server/src/main/java/org/opensearch/plugins/IdentityPlugin.java

```diff
@@ -34,6 +40,15 @@ public interface IdentityPlugin {
      */
     TokenManager getTokenManager();
 
+    /**
+     * Returns a function used to wrap each rest request before handling the request.
+     * The returned {@link UnaryOperator} is called for every incoming rest request and receives
+     * the original rest handler as it's input.
+     */
+    default UnaryOperator<RestHandler> authenticate(ThreadContext threadContext) {
```

Comment:
@cwperks I haven't look into the full change yet, but I think this is not the direction we should be taken:
 - the authentication is 100% the business of the identity plugin implementors
 - the core provide hooks (request / response processing) for identity plugin to plug into, if are missing something here, please share
 - the only thing the identity plugin should provide is the `Subject` (current caller identity), how it is being established is solely the internals of the identity plugin implementations

Reply:
> the core provide hooks (request / response processing) for identity plugin to plug into, if we are missing something here, please share

That's exactly what this PR is. Its a hook so that an IdentityPlugin can provide the implemenation.

What this PR does is mimic the existing extension point that Security uses which is [ActionPlugin.getRestHandlerWrapper](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L130-L153). This is the same extension point, but moved into IdentityPlugin which I think is where it really belongs.

Reply:
[ActionPlugin.getRestHandlerWrapper](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L130-L153) is the primary extension point that Security uses to extract and authenticate the authinfo from the request (usually the AUTHORIZATION header, but can be a client cert or other http headers). If the request cannot be authenticated then security will not delegate to the original rest handler and responds UNAUTHORIZED back to the caller.

Reply:
> That's exactly what this PR is. Its a hook so that an IdentityPlugin can provide the implemenation.

It shouldn't expose anything beyond the `Subject`

Reply:
Got it. My ultimate goal is to remove the identity feature flag before asking [plugins to adopt the IdentityAwarePlugin](https://github.com/opensearch-project/opensearch-plugins/issues/238) that we introduced in the last PR. If the feature flag is not removed, then plugin teams need to run integ tests with the identity feature flag enabled. 

The problem is the current code in the RestController [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/RestController.java#L591-L601). That code assumes that all auth tokens are present in the AUTHORIZATION header which does not encompass all of the ways that auth info can be included on a Rest Request.

Since the current ActionPlugin.getRestHandlerWrapper is battle tested, I would like to keep that mechanism for auth. I see 2 possible approaches:

1. Change the UserSubject interface to accept a more generic LoginContext which can include the entire RestRequest. Ideally it includes the channel too (since Security does eagerly respond on the channel), but we can keep the existing interface that returns a bool of the auth result. It would be a decent effort on the security plugin side, but possible.
2. Remove `handleAuthenticateUser` from the RestController

Reply:
@stephen-crawford wdyt about the idea of removing the existing code behind the identity feature flag within the RestController? The Shiro identity plugin can adopt a similar approach to the security plugin where it can implement `ActionPlugin.getRestHandlerWrapper()` to determine the current subject from a rest request.

Reply:
The issue with the current logic in the RestController is the rigidity. Its not necessarily wrong its just that to extend it to work with the security plugin will require significant work and risk.

The [recent change to provide a replacement for stashContext](https://github.com/opensearch-project/OpenSearch/pull/14630) is gated behind the Identity feature flag.

The immediate goal is to get rid of the identity feature flag, but keep IdentityPlugin marked as experimental which means that it is subject to change.

@reta brought up a point about leaving it up to the IdentityPlugin to figure out how to hydrate the current subject and there are existing extension points that an IdentityPlugin can plugin into to get the information necessary to determine the current subject

Reply:
Hmm I see. Maybe I was not understanding Reta's initial comments. 

I obviously am in favor of whatever has first the least risk and then less work second. But if this approach has both seems like a no brainer. 

Let me review and see where things are at.

Reply:
Marking thread as resolved. This code has been removed.